[MRG after #12145] Add "Randomized SVD" solver option to KernelPCA for faster partial decompositions, like in PCA (#12069)
scikit-learn/scikit-learn: doc/modules/decomposition.rst

      Halko, et al., 2009

> .. _kernel_PCA:

Reviewer: Any specific reason to move this below?

Author: Yes I think that kernel PCA deserves higher visibility, in other words it should appear in the toc that is generated on the left of the page (as of now you see only PCA). It makes navigation a bit clearer to separate all things relative to PCA, from all things related to kPCA

Author: For an easier review:

 - Current doc: https://scikit-learn.org/stable/modules/decomposition.html (no kPCA in the toc on the left, this is in an obscure subsection)
 - New proposal: kPCA is easy to spot https://100494-843222-gh.circle-artifacts.com/0/doc/modules/decomposition.html

For more details on how to use Sparse PCA, see the Examples section, below.

> .. |spca_img| image:: ../auto_examples/decomposition/images/sphx_glr_plot_faces_decomposition_006.png

Author: ```suggestion
.. |spca_img| image:: ../auto_examples/decomposition/images/sphx_glr_plot_faces_decomposition_005.png
```

     R. Jenatton, G. Obozinski, F. Bach, 2009

.. _kernel_PCA:

Kernel Principal Component Analysis (kPCA)
==========================================

Exact Kernel PCA
----------------

:class:`KernelPCA` is an extension of PCA which achieves non-linear
dimensionality reduction through the use of kernels (see :ref:`metrics`). It
has many applications including denoising, compression and structured
prediction (kernel dependency estimation). :class:`KernelPCA` supports both
``transform`` and ``inverse_transform``.

.. figure:: ../auto_examples/decomposition/images/sphx_glr_plot_kernel_pca_001.png
    :target: ../auto_examples/decomposition/plot_kernel_pca.html
    :align: center
    :scale: 75%

.. topic:: Examples:

    * :ref:`sphx_glr_auto_examples_decomposition_plot_kernel_pca.py`

.. _kPCA_Solvers:

Choice of solver for Kernel PCA
-------------------------------

> In :class:`KernelPCA`, the number of components found is equal to the number of

Reviewer: May I suggest

> While in PCA the number of components is bounded by the number of features, in KPCA the number of components is bounded by the number of samples.

Author: Great proposal!

samples. Many real-world datasets have large number of samples! In these cases
finding *all* the components with a full kPCA is a waste of computation time,
as data is mostly described by the first few components
(e.g. ``n_components<=100``).

The optional parameter ``eigen_solver='randomized'`` can be used to
*significantly* reduce the computation time when the number of requested
``n_components`` is small compared with the number of samples.
This is done automatically by default if you do not specify any
``eigen_solver`` and if the required ``n_components`` is less than 80% of
the number of samples, see :class:`KernelPCA` for details.

For example the figure below shows the same "circles" dataset (top left)
reconstructed with various approximations (number of components decreases from
2000 - full - to 4). You can compare the execution times vs. the reconstruction
quality :

> .. figure:: ../auto_examples/decomposition/images/sphx_glr_plot_kernel_pca_approximate_001.png

Reviewer: I'd remove honestly.

The timings there are only relevant for the machine where the benchmark was run on, and the figure doesn't bring any specific insight since we alreaady have the one above. All the plots on the first row are indistinguishable, same for the second row.

Reviewer: Also this is pretty much redundant with the complexity analysis below (thanks for that)

Author: ok, I'll remove it.

    :target: ../auto_examples/decomposition/plot_kernel_pca_approximate.html
    :align: center
    :scale: 75%

The time complexity of the randomized :class:`KernelPCA` is
:math:`O(n_{\mathrm{samples}}^2 \cdot n_{\mathrm{components}})`
instead of :math:`O(n_{\mathrm{samples}}^3)` for the exact method
implemented with ``eigen_solver='dense'``.

The memory footprint of randomized :class:`KernelPCA` is also proportional to
:math:`2 \cdot n_{\mathrm{samples}} \cdot n_{\mathrm{components}}` instead of
:math:`n_{\mathrm{samples}}^2` for the exact method.

.. topic:: Examples:

    * :ref:`sphx_glr_auto_examples_decomposition_plot_kernel_pca_approximate.py`

> Note: this technique is the same than in :ref:`RandomizedPCA`.

Reviewer: same as

And please put this above the example section

Author: done

     R. Jenatton, G. Obozinski, F. Bach, 2009

.. _kernel_PCA:

Kernel Principal Component Analysis (kPCA)
==========================================

Exact Kernel PCA
----------------

:class:`KernelPCA` is an extension of PCA which achieves non-linear
dimensionality reduction through the use of kernels (see :ref:`metrics`). It
has many applications including denoising, compression and structured
prediction (kernel dependency estimation). :class:`KernelPCA` supports both
``transform`` and ``inverse_transform``.

.. figure:: ../auto_examples/decomposition/images/sphx_glr_plot_kernel_pca_001.png
    :target: ../auto_examples/decomposition/plot_kernel_pca.html
    :align: center
    :scale: 75%

.. topic:: Examples:

    * :ref:`sphx_glr_auto_examples_decomposition_plot_kernel_pca.py`

.. _Randomized_kPCA:

> Kernel PCA using randomized SVD

Reviewer: Call this "Choice of solver"?

It'd be nice to also briefly discuss the other ones, but don't feel obliged

Author: Good proposition, I'll work on that tomorrow

Author: Actually in PCA it is named like this. https://scikit-learn.org/stable/modules/decomposition.html#pca-using-randomized-svd

But I agree that your proposal makes more sense, so I'll update the title still.

Author: Done. I discuss both randomized and dense, but not arpack as I do not know it well enough.

-------------------------------

> In :ref:`kernel_PCA`, the number of components found is equal to the number of

Reviewer: ```suggestion
In :class:`KernelPCA`, the number of components found is equal to the number of
```
Otherwise you're just linking to the section above

Author: Indeed, good catch

> In :ref:`kernel_PCA`, the number of components found is equal to the number of

Reviewer: I don't understand this sentence: what is the number of components found?

Also I don't think that's True. WIth n_components=None, a (100, 10) matrix will be transformed to a matrix of the same shape

Author: The number of principal components (PC) is the number of eigenvalues/eigenvectors found ; for Kernel PCA they live in the kernel space if I remember well, not in the original space. If you set `n_components=None` and throw in a (100, 10) matrix, 100 components will be found (see `model._lambdas`)

> samples. Many real-world datasets have large number of samples ! In these cases

Reviewer: ```suggestion
samples. Many real-world datasets have large number of samples! In these cases
```

or 

```suggestion
samples. Many real-world datasets have large number of samples. In these cases
```

finding *all* the components with a full kPCA is a waste of computation time,
as data is mostly described by the first few components

> (``n_components<=100``).

Reviewer: add "e.g." ? I don't think we can generalize this to all datasets

Author: Indeed.

```suggestion
(e.g. ``n_components<=100``).
```

The optional parameter ``eigen_solver='randomized'`` can be used to

> *drastically* reduce the computation time when the number of requested

Reviewer: ```suggestion
reduce the computation time when the number of requested
```
benchmarks suggest it's not that drastic I would say, i.e. not multiple orders of magnitude faster

Author: I would need confirmation from an expert, but it should be O(`n_components`/`n_samples`) faster. I call this "drastically" when n comps is < 100 and n_samples is > 10000... But it is a matter of language

Reviewer: Use "significantly"? I don't think we should bike shed on this...

For info, this is an entry of 0.23 

- |Fix| |Efficiency| :class:`linear_model.ARDRegression` is more stable and
  much faster when `n_samples > n_features`

where "much faster" actually means "you couldn't fit more than 1k samples before but now this scales up to more than 500k samples" in 1 sec

Author: sure, I'll use significantly. I learned [a new expression](https://www.urbandictionary.com/define.php?term=bikeshed), thanks !

``n_components`` is small compared with the number of samples.
This is done automatically by default if you do not specify any
``eigen_solver`` and if ``n_components`` is less than 80% of
the number of samples, see :class:`KernelPCA` for details.

For example the figure below shows the same "circles" dataset (top left)
reconstructed with various approximations (number of components decreases from
2000 - full - to 4). You can compare the execution times vs. the reconstruction
quality :

.. figure:: ../auto_examples/decomposition/images/sphx_glr_plot_kernel_pca_approximate_001.png
    :target: ../auto_examples/decomposition/plot_kernel_pca_approximate.html
    :align: center
    :scale: 75%

.. topic:: Examples:

    * :ref:`sphx_glr_auto_examples_decomposition_plot_kernel_pca_approximate.py`

Note: this technique is the same than in :ref:`RandomizedPCA`.

Kernel PCA solvers benchmarks

> -----------------------------

Reviewer: We usually don't include benchmarks in the UG. Guidelines as you did above are good enough

Author: It is a bit sad, as these benchmarks would then exist but noone would know about the code being there. How can we make this at least visible so that people can quickly find the code if needed ?

Reviewer: benchmarks will get more visibility once we have a dedicated benchmark suite, but for now these should be kept out of the UG.

Author: ok, I'll remove them.

     R. Jenatton, G. Obozinski, F. Bach, 2009

.. _kernel_PCA:

> Kernel Principal Component Analysis (kPCA)

Author: This acronym is widely popular - I prefer to leave it for search engines and people knowing only the acronym

> Kernel Principal Component Analysis (kPCA)

Reviewer 2: Sounds good.

Choice of solver for Kernel PCA
-------------------------------

While in :class:`PCA` the number of components is bounded by the number of
features, in :class:`KernelPCA` the number of components is bounded by the
number of samples. Many real-world datasets have large number of samples! In
these cases finding *all* the components with a full kPCA is a waste of
computation time, as data is mostly described by the first few components
(e.g. ``n_components<=100``).

The optional parameter ``eigen_solver='randomized'`` can be used to
*significantly* reduce the computation time when the number of requested
``n_components`` is small compared with the number of samples.
This is done automatically by default if you do not specify any
``eigen_solver`` and if the required ``n_components`` is less than 80% of
the number of samples, see :class:`KernelPCA` for details.

The time complexity of the randomized :class:`KernelPCA` is
:math:`O(n_{\mathrm{samples}}^2 \cdot n_{\mathrm{components}})`
instead of :math:`O(n_{\mathrm{samples}}^3)` for the exact method
implemented with ``eigen_solver='dense'``.

The memory footprint of randomized :class:`KernelPCA` is also proportional to

> :math:`2 \cdot n_{\mathrm{samples}} \cdot n_{\mathrm{components}}` instead of

Reviewer 3: Very nit: Do we include the `2` here when comparing this to `n_samples^2`?

Author: Frankly I can not answer easily here, the details are quite far away now :)

Author: Now I remember: this is a copy of the documentation in the [randomized PCA](https://scikit-learn.org/stable/modules/decomposition.html#pca-using-randomized-svd) that is already in the sklearn doc:

![image](https://user-images.githubusercontent.com/3236794/99870462-59d29980-2bd3-11eb-82dd-7cb4757b0e7d.png)

Author: But honestly I am not sure (even for this original comment in PCA doc, I don't know who wrote it ) that this comment is relative to usage of the "randomized" solver, or simply to the fact that when `n_components` is less than the max, then only this number of support vectors have to be kept in memory. We should maybe contact the original author of the PCA doc to understand why he/she wrote this ? As for KernelPCA, I'll just align.
